auth/server/src/main/java/com/seagates3/policy/PolicyUtil.java

```diff
@@ -1,229 +1,258 @@
+   * @return
+   */
+ public
+  static List<String> getAllMatchingActions(String inputAction) {
+    List<String> matchingActions = new ArrayList<>();
+    for (String action : S3Actions.getInstance()
+             .getBucketOperations()
+             .keySet()) {
+      if (isPatternMatching(action.toLowerCase(), inputAction.toLowerCase())) {
+        matchingActions.add(action);
+      }
+    }
+    for (String action : S3Actions.getInstance()
+             .getObjectOperations()
+             .keySet()) {
+      if (isPatternMatching(action.toLowerCase(), inputAction.toLowerCase())) {
+        matchingActions.add(action);
+      }
+    }
+    for (String action : IAMActions.getInstance().getOperations().keySet()) {
+      if (isPatternMatching(action.toLowerCase(), inputAction.toLowerCase())) {
+        matchingActions.add(action);
+      }
+    }
+
+    return matchingActions;
+  }
+
+  /**
+   * Function that matches input string with given wildcard pattern
+   *
+   * @param input - Source string to match the pattern with
+   * @param pattern - Pattern with wildchar characters to match. eg- GetObj*
+   *or GetObjectAc?
+   * @return - true/false
+   */
+ public
+  static boolean isPatternMatching(String input, String pattern) {
+    if (input == null) return false;
+    int inputLength = input.length();
+    int patternLength = pattern.length();
+    if (patternLength == 0) return (inputLength == 0);
+    boolean[][] data = new boolean[inputLength + 1][patternLength + 1];
+    for (int count = 0; count < inputLength + 1; count++)
+      Arrays.fill(data[count], false);
+    data[0][0] = true;
+    // Only '*' can match with empty string
+    for (int count = 1; count <= patternLength; count++)
+      if (pattern.charAt(count - 1) == '*') data[0][count] = data[0][count - 1];
+    for (int i = 1; i <= inputLength; i++) {
+      for (int j = 1; j <= patternLength; j++) {
+        if (pattern.charAt(j - 1) == '*')
+          data[i][j] = data[i][j - 1] || data[i - 1][j];
+        else if (pattern.charAt(j - 1) == '?' ||
+                 input.charAt(i - 1) == pattern.charAt(j - 1))
+          data[i][j] = data[i - 1][j - 1];
+        else
+          data[i][j] = false;
+      }
+    }
+    return data[inputLength][patternLength];
+  }
+
+ public
+  static boolean isActionValidForResource(List<String> actions,
+                                          int slashPosition) {
+    boolean isValid = false;
+    for (String action : actions) {
+      if ((S3Actions.getInstance().getBucketOperations().keySet().contains(
+              action)) &&
+          slashPosition == -1) {  // Bucket operation and object not present
+        isValid = true;
+        break;
+      }
+      if ((S3Actions.getInstance().getObjectOperations().keySet().contains(
+              action)) &&
+          slashPosition != -1) {  // object operation and object is present
+        isValid = true;
+        break;
+      }
+    }
+    return isValid;
+  }
+
+ public
```

Comment:
this is a very generic level utility function. Do we have a generic utility file where this can be moved?

Reply:
We dont have any generic Util class for all classes. As we using it for Policy, moved the method into PolicyUtil